### PR TITLE
[flutter_tools] use frontend_server for web test compilation

### LINKED ISSUE #44583: Windows 10: it is confirmed that "flutter test --platform chrome" does not work
Running ```flutter test --platform chrome -v```
on new flutter project (created by ```flutter create temp```)
raised following error (both ```flutter test``` and ```flutter run -d chrome``` work well):

<details>
  <summary>logs</summary>

```
c:\wikibulary\dart\algorithm\temp>flutter test --platform chrome -v
[  +43 ms] executing: [c:\flutter\] git -c log.showSignature=false log -n 1 --pretty=format:%H
[ +148 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +12 ms] fbabb264e0ab3e090d6ec056e0744aaeb1586735
[   +3 ms] executing: [c:\flutter\] git describe --match v*.*.* --first-parent --long --tags
[ +156 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[   +4 ms] v1.10.15-0-gfbabb264e
[  +63 ms] executing: [c:\flutter\] git rev-parse --abbrev-ref --symbolic @{u}
[  +95 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[   +4 ms] origin/dev
[   +2 ms] executing: [c:\flutter\] git ls-remote --get-url origin
[  +82 ms] Exit code 0 from: git ls-remote --get-url origin
[   +4 ms] https://github.com/flutter/flutter.git
[ +189 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[   +6 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[   +2 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[   +2 ms] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[  +16 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[   +7 ms] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[   +5 ms] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[+16009 ms] Starting daemon...
[ +258 ms] Initializing inputs
[  +29 ms] Reading cached asset graph...
[ +369 ms] Reading cached asset graph completed, took 363ms

[ +530 ms] Checking for updates since last build...
[ +804 ms] Checking for updates since last build completed, took 804ms

[  +19 ms] Initializing inputs
[   +6 ms] Reading cached asset graph...
[   +5 ms] Reading cached asset graph completed, took 363ms
[   +2 ms] Checking for updates since last build...
[   +1 ms] Checking for updates since last build completed, took 804ms
[+1837 ms] About to build [web, test]...
[  +20 ms] Running build...
[   +2 ms] Running build completed, took 127ms

[  +10 ms] Caching finalized dependency graph...
[   +9 ms] Caching finalized dependency graph completed, took 190ms

[   +4 ms] Succeeded after 336ms with 0 outputs (0 actions)

00:00 +0: compiling c:\wikibulary\dart\algorithm\temp\test\widget_test.dart                                                                                                [+20798 ms] Serving tests at http://localhost:61718/static/index.html?managerUrl=ws%3A%2F%2Flocalhost%3A61718%2F0&debug=false
[ +925 ms] Did not find anything for request: widget_test.dart.browser_test.dart.js
00:01 +0 -1: compiling c:\wikibulary\dart\algorithm\temp\test\widget_test.dart [E]
  Failed to load "c:\wikibulary\dart\algorithm\temp\test\widget_test.dart": Failed to load script at http://localhost:61718/widget_test.dart.browser_test.dart.js.
00:01 +0 -1: Some tests failed.

#0      throwToolExit (package:flutter_tools/src/base/common.dart:28:3)
#1      TestCommand.runCommand (package:flutter_tools/src/commands/test.dart:258:7)
<asynchronous suspension>
#2      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:157:29)
<asynchronous suspension>
#3      _rootRun (dart:async/zone.dart:1124:13)
#4      _CustomZone.run (dart:async/zone.dart:1021:19)
#5      _runZoned (dart:async/zone.dart:1516:10)
#6      runZoned (dart:async/zone.dart:1463:12)
#7      AppContext.run (package:flutter_tools/src/base/context.dart:156:18)
<asynchronous suspension>
#8      FastFlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:750:20)
#9      CommandRunner.runCommand (package:args/command_runner.dart:197:27)
<asynchronous suspension>
#10     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:416:21)
<asynchronous suspension>
#11     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:157:29)
<asynchronous suspension>
#12     _rootRun (dart:async/zone.dart:1124:13)
#13     _CustomZone.run (dart:async/zone.dart:1021:19)
#14     _runZoned (dart:async/zone.dart:1516:10)
#15     runZoned (dart:async/zone.dart:1463:12)
#16     AppContext.run (package:flutter_tools/src/base/context.dart:156:18)
<asynchronous suspension>
#17     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:367:19)
<asynchronous suspension>
#18     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
#19     new Future.sync (dart:async/future.dart:222:31)
#20     CommandRunner.run (package:args/command_runner.dart:112:14)
#21     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:251:18)
#22     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:63:22)
<asynchronous suspension>
#23     _rootRun (dart:async/zone.dart:1124:13)
#24     _CustomZone.run (dart:async/zone.dart:1021:19)
#25     _runZoned (dart:async/zone.dart:1516:10)
#26     runZoned (dart:async/zone.dart:1500:12)
#27     run.<anonymous closure> (package:flutter_tools/runner.dart:61:18)
<asynchronous suspension>
#28     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:157:29)
<asynchronous suspension>
#29     _rootRun (dart:async/zone.dart:1124:13)
#30     _CustomZone.run (dart:async/zone.dart:1021:19)
#31     _runZoned (dart:async/zone.dart:1516:10)
#32     runZoned (dart:async/zone.dart:1463:12)
#33     AppContext.run (package:flutter_tools/src/base/context.dart:156:18)
<asynchronous suspension>
#34     runInContext (package:flutter_tools/src/context_runner.dart:64:24)
<asynchronous suspension>
#35     run (package:flutter_tools/runner.dart:50:10)
#36     main (package:flutter_tools/executable.dart:65:9)
<asynchronous suspension>
#37     main (file:///c:/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#38     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:303:32)
#39     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)

```

</details>

flutter doctor:

```
[√] Flutter (Channel dev, v1.10.15, on Microsoft Windows [Version 10.0.18362.418], locale en-US)
    • Flutter version 1.10.15 at c:\flutter
    • Framework revision fbabb264e0 (10 days ago), 2019-11-02 01:36:07 +0300
    • Engine revision 8ea19b1c76
    • Dart version 2.6.0 (build 2.6.0-dev.8.2 bbe2ac28c9)

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\pzika\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    X No Java Development Kit (JDK) found; You must have the environment variable JAVA_HOME set and the java binary in
      your PATH. You can download the JDK from https://www.oracle.com/technetwork/java/javase/downloads/.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[!] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    X Unable to determine bundled Java version.
    • Try updating or re-installing Android Studio.

[√] VS Code (version 1.40.0)
    • VS Code at C:\Users\pzika\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.6.0

[√] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 78.0.3904.97
    • Web Server • web-server • web-javascript • Flutter Tools

! Doctor found issues in 2 categories.
```

### LINKED ISSUE #50594: Remove chrome platform from `flutter test`
Early in the web support phase, I added support for directly testing in a chrome browser for a simulated test environment. This may have proved useful in helping the web bootstrap, but long-term it does not present a very compelling test environment. Since it doesn't actually perform rendering/painting it only really tests the interface between the engine and framework. Compared to actual integration tests on real browsers, it is of limited use for ensuring correctness. In addition, these tests are highly flaky on CI due to in part to the orchestration complexity.

Unfortunately it also adds a lot of dependencies to the flutter tool. This was acceptable since we needed them anyway for `flutter run`. When we move to the frontend_server, many of these dependencies are no longer needed though. Notably, the build_runner dependency is one of two transitive mirrors dependencies that prevents an AOT built tool

I propose that in the short-term, we remove support for `flutter test --platform=chrome` from the tool into a special purpose command in `dev/bots/`. This would still be usable for the framework CI, but would no longer be accessible for the tool.

@nturgut @yjbanov @zanderso 

EDIT:

@yjbanov @Piinks, are there any implications for web goldens?

@mehmetf @chingjun, is this code used in google3 at all?

### LINKED ISSUE #52685: framework stack trace/frame tests don't cover web release/debug mode
Specifically, since `flutter test -p chrome` still uses build runner, the actual stack frames someone will see when using `flutter run` is different.

Finally, these tests do not currently cover profile/release mode

### LINKED ISSUE #61120: Remove usage of build_runner for flutter web framework unit test
We do not want to require build_runner and build_web_compilers to be updated to support sound null safety before we migrate the tests. Additionally, I would like to remove `flutter test --platform=chrome` altogether, as it is not currently debuggable nor easy to use, which will require our tests to have a different setup.

Ideally this new workflow would support debugging via dwds.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

* Uses same compiler for testing that is used for development
* Unblocks running web tests in sound mode
* Removes last build_runner usage.

As a follow up, there is more refactoring required so that the bootstrap logic is shared between web run and test

Fixes https://github.com/flutter/flutter/issues/61120
Fixes https://github.com/flutter/flutter/issues/52685
Fixes https://github.com/flutter/flutter/issues/50594
Fixes https://github.com/flutter/flutter/issues/50594
Fixes https://github.com/flutter/flutter/issues/44583

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
TestOn chrome is not supported anymore, so we'll need to add the failing tests to the skip list

**Comment by skia-gold** [comment]
Gold has detected about 211 untriaged digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/70714

**Comment by skia-gold** [comment]
Gold has detected about 246 untriaged digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/70714

File: packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart [hunk 1]
```diff
@@ -2,266 +2,257 @@
...
+      ..createSync(recursive: true);
+    final List<File> generatedFiles = <File>[];
+    for (final String testFilePath in testFiles) {
+      final List<String> relativeTestSegments = globals.fs.path.split(
+        globals.fs.path.relative(testFilePath, from: projectDirectory.childDirectory('test').path));
+      final File generatedFile = globals.fs.file(
+        globals.fs.path.join(outputDirectory.path, '${relativeTestSegments.join('_')}.test.dart'));
+      generatedFile
+        ..createSync(recursive: true)
+        ..writeAsStringSync(_generateEntrypoint(relativeTestSegments.join('/'), testFilePath));
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart:48]: Would it be noticeably faster to fan out these writes into multiple concurrent writes and wait on them using `Future.wait`?

File: packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart [hunk 2]
```diff
@@ -2,266 +2,257 @@
...
+      generatedFiles.add(generatedFile);
+    }
+    final StringBuffer buffer = StringBuffer('// @dart=2.8\n');
+    for (final File generatedFile in generatedFiles) {
+      buffer.writeln('import "${globals.fs.path.basename(generatedFile.path)}";');
+    }
+    buffer.writeln('void main() {}');
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart:55]: Can you please leave a comment explaining why the main.dart file has an empty main function?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart:57]: Done

File: packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart:48]: This part of the process doesn't take much time - performance optimization is not a goal of this PR

File: packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart [hunk 3]
```diff
@@ -2,266 +2,259 @@
...
+import '../bundle.dart';
+import '../compile.dart';
+import '../convert.dart';
 import '../globals.dart' as globals;
-import '../platform_plugins.dart';
-import '../plugins.dart';
-import '../project.dart';
 import '../web/compile.dart';
 
 /// A build_runner specific implementation of the [WebCompilationProxy].
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart:20]: Does this comment need an update?

**Review by yjbanov — APPROVED** [review]

**Review by zanderso — COMMENTED** [review]
`web_compilation_delegate.dart` looks very fiddly. I suspect that file would benefit from a paragraph or two at the top explaining how it's all put together.

**Comment by jonahwilliams** [comment]
web_compilation_delegate.dart needs substantial refactoring/removing - but thought the line length doesn't show it this was a monster of a PR to get working, so I tried to keep most of the APIs the same. I have a lot of follow up work to do, but that should be much easier to get started on

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/isolated/web_compilation_delegate.dart]: Done

File: dev/bots/test.dart [hunk 4]
```diff
@@ -782,7 +795,7 @@ Future<void> _runWebUnitTests() async {
     )
     .whereType<File>()
     .map<String>((File file) => path.relative(file.path, from: flutterPackageDirectory.path))
-    .where((String filePath) => !kWebTestFileKnownFailures.contains(filePath))
+    .where((String filePath) => !kWebTestFileKnownFailures.contains(path.split(filePath).join('/')))
```
Comment (jonahwilliams) [inline dev/bots/test.dart:798]: allows running on windows

File: packages/flutter_tools/lib/src/isolated/devfs_web.dart [hunk 5]
```diff
@@ -942,18 +942,12 @@ class WebDevFS implements DevFS {
     File metadataFile;
     List<String> modules;
     try {
-      final Directory parentDirectory =
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/isolated/devfs_web.dart:945]: Are you reformatting 80 character lines to longer?  We're just fighting each other's autoformatting settings.  We should get on the same page so we're not thrashing.

File: packages/flutter/test/foundation/defines_test.dart [hunk 6]
```diff
@@ -1,14 +0,0 @@
-// Copyright 2014 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/test/foundation/defines_test.dart:1]: This is still unresolved: https://github.com/dart-lang/sdk/issues/41764

File: packages/flutter_tools/lib/src/isolated/devfs_web.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/isolated/devfs_web.dart:945]: I don't generally mind if you reformat things to be shorter, but that has to be a reciprocal agreement

Comment (jmagman) [reply packages/flutter_tools/lib/src/isolated/devfs_web.dart:945]: You know I wouldn't have any character limits if I were Boss.  But I would rather just hit a button to format.  We can talk about it outside this PR, but I don't think you even touched this file other than formatting it?

**Pull request merged**